### PR TITLE
Rework & refactor programmer module to integrate hardware manager for cable and device resources consolidation

### PR DESCRIPTION
**Pull request by guanyung-rs**
> ### Motivate of the pull request
> - [ ] To address an existing issue. If so, please provide a link to the issue: <issue id>
> - [x] Breaking new feature. If so, please describe details in the description part.

> ### Describe the technical details
> #### What is currently done? (Provide issue link if applicable)
The rational of this PR:
Previously, the programmer not only doing the programming feature, it also manages the cable and device resources. This PR rework & refactor programmer module and integrate Hardware Manager module  to separate out the responsibilities of each module. Programmer module is doing the flash programming and FPGA configuration. The Hardware Manager manages the cable and device resources. The openOCD adapter contains all the logic to communicate with openocd executable via command invocation. 

- Implemented jtag_frequency tcl command to support jtag speed configuration
- Added Hardware Manager class to manage cable and device resources discovery algorithm. This class is now fully responsible for handling cable and device information in the system. It eliminates the previous implementation.
- Implement OpenocdAdatper class where this class implements the JtagAdapter and ProgrammingAdapter interface functions - `scan, program_fpga, program_flash, program_otp, query_fpga_status`. The implementation is done via OpenOCD execution commands.
- Added JtagAdapter and ProgrammingAdapter interfaces
- Added ProgrammingTool class as a thin wrapper for programming functions.
- Split Tap, Cable and Device classes to independent header file
- Add unit tests code to add test coverage
- Clean up error code 
- Remove and clean up Programmer_helper cpp
- Remove unused unit testing code and others stale code
- Tested the programmer-gui operation
- 
Screenshot 2023-12-27 105750
Screenshot 2023-12-27 105924

> 
> 
> 
> 
> 
>
> #### What does this pull request change?
> 
> 
> 
> 
> 

> ### Which part of the code base require a change
> 
> - [x] Library: Programmer and Hardware Manager
> - [ ] Plug-in: <Specify the plugin name>
> - [ ] Engine
> - [ ] Documentation
> - [ ] Regression tests
> - [ ] Continous Integration (CI) scripts

> ### Impact of the pull request

> - [ ] Require a change on Quality of Results (QoR)
> - [ ] Break back-compatibility. If so, please list who may be influenced.

### CODE REVIEW

File: src/Configuration/CFGCommon/CFGArg.json [hunk 1]
```diff
@@ -248,7 +265,9 @@
...
         "  programmer list_device",
         "To list all connected programming cable:",
-        "  programmer list_cable"
+        "  programmer list_cable",
+        "To configure JTAG frequency for specified cable:",
+        "  programmer jtag_frequency -c <cable_index or cable_name> <frequency in kHz>"
     ]}
   ]
-}
+}
```
Comment (chernyee-rs) [inline src/Configuration/CFGCommon/CFGArg.json:273]: add empty line

File: src/Configuration/HardwareManager/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    DESTINATION ${CMAKE_INSTALL_LIBDIR}/foedag
+  )
+endforeach()
+
+foreach(file ${SRC_H_INSTALL_LIST})
+  install(
+    FILES ${PROJECT_SOURCE_DIR}/${file}
+    DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/foedag/HardwareManager
+  )
+endforeach()
```
Comment (chernyee-rs) [inline src/Configuration/HardwareManager/CMakeLists.txt:89]: add empty line

File: src/Configuration/HardwareManager/Cable.h [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+struct CompareCable {
+  bool operator()(const Cable& a, const Cable& b) const {
+    if (a.name < b.name) return true;
+    if (a.name > b.name) return false;
+    return a.index < b.index;
+  }
+};
+
+}  // namespace FOEDAG
+#endif  // __CABLE_H__
```
Comment (chernyee-rs) [inline src/Configuration/HardwareManager/Cable.h:56]: add empty line

File: src/Configuration/HardwareManager/Device.h [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  std::string name;
+  int flashSize;
+  DeviceType type;
+  Cable cable;
+  Tap tap;
+};
+
+}  // namespace FOEDAG
+
+#endif  // __DEVICE_H__
```
Comment (chernyee-rs) [inline src/Configuration/HardwareManager/Device.h:45]: add empty line

File: src/Configuration/HardwareManager/HardwareManager.cpp [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+
+  return devices;
+}
+
+const std::vector<HardwareManager_DEVICE_INFO>&
+HardwareManager::get_device_db() {
+  return m_device_db;
+}
+
+}  // namespace FOEDAG
```
Comment (chernyee-rs) [inline src/Configuration/HardwareManager/HardwareManager.cpp:280]: add empty line

File: src/Configuration/HardwareManager/HardwareManager.h [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+  static const std::vector<HardwareManager_DEVICE_INFO> &get_device_db();
+
+ private:
+  static const std::vector<HardwareManager_CABLE_INFO> m_cable_db;
+  static const std::vector<HardwareManager_DEVICE_INFO> m_device_db;
+  JtagAdapter *m_adapter;
+};
+
+}  // namespace FOEDAG
+#endif  //__HARDWAREMANAGER_H__
```
Comment (chernyee-rs) [inline src/Configuration/HardwareManager/HardwareManager.h:81]: add empty line

File: src/Configuration/HardwareManager/JtagAdapter.h [hunk 7]
```diff
@@ -17,32 +17,19 @@ GNU General Public License for more details.
...
- private:
-  Cable m_cable;
-  std::vector<Device> m_devices;
+  virtual std::vector<uint32_t> scan(const Cable &cable) = 0;
 };
-
-}  // namespace FOEDAG
\ No newline at end of file
+}  // namespace FOEDAG
+#endif  //__JTAGADAPTER_H__
```
Comment (chernyee-rs) [inline src/Configuration/HardwareManager/JtagAdapter.h:35]: add empty line

File: src/Configuration/HardwareManager/OpenocdAdapter.cpp [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+  ss << " -c \"" << cmd << "\"";
+  ss << " -c \"exit\"";
+
+  // run the command
+  int res = CFG_execute_cmd("OPENOCD_DEBUG_LEVEL=-3 " + m_openocd + ss.str(),
+                            output, nullptr, stop);
+  return res;
+}
+
+}  // namespace FOEDAG
```
Comment (chernyee-rs) [inline src/Configuration/HardwareManager/OpenocdAdapter.cpp:380]: add empty line

File: src/Configuration/HardwareManager/OpenocdAdapter.h [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+ private:
+  int execute(const Cable& cable, std::string cmd, std::string& output);
+  std::string m_openocd;
+  std::vector<Tap> m_taplist;
+  std::string m_last_output;
+};
+
+}  // namespace FOEDAG
+
+#endif  //__OPENOCDADAPTER_H__
```
Comment (chernyee-rs) [inline src/Configuration/HardwareManager/OpenocdAdapter.h:90]: add empty line

File: src/Configuration/HardwareManager/OpenocdHelper.cpp [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+  ss << " -c \"init\"";
+  ss << " -c \"" << cmd << "\"";
+  ss << " -c \"exit\"";
+
+  std::string openocd_command = "OPENOCD_DEBUG_LEVEL=-3 " + openocd + ss.str();
+
+  return openocd_command;
+}
+
+}  // end namespace FOEDAG
```
Comment (chernyee-rs) [inline src/Configuration/HardwareManager/OpenocdHelper.cpp:124]: add empty line

File: src/Configuration/HardwareManager/OpenocdHelper.h [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+std::string build_tap_config(const std::vector<Tap>& taplist);
+std::string build_target_config(const Device& device);
+std::string create_openocd_command(const std::string& subcmd,
+                                   const Device& device,
+                                   const std::vector<Tap>& taplist,
+                                   const std::string& bitfile,
+                                   const std::string& openocd);
+}  // namespace FOEDAG
+
+#endif  // __OPENOCDHELPER_H__
```
Comment (chernyee-rs) [inline src/Configuration/HardwareManager/OpenocdHelper.h:43]: add empty line

File: src/Configuration/HardwareManager/ProgrammingAdapter.h [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+                          std::ostream* outStream = nullptr,
+                          OutputMessageCallback callbackMsg = nullptr,
+                          ProgressCallback callbackProgress = nullptr) = 0;
+  virtual int query_fpga_status(const Device& device, CfgStatus& cfgStatus,
+                                std::string& outputMessage) = 0;
+};
+
+}  // namespace FOEDAG
+
+#endif  // PROGRAMMINGADAPTER_H
```
Comment (chernyee-rs) [inline src/Configuration/HardwareManager/ProgrammingAdapter.h:42]: add empty line

File: src/Configuration/HardwareManager/Tap.h [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+namespace FOEDAG {
+struct Tap {
+  uint32_t index;
+  uint32_t idcode;
+  uint32_t irlength;
+};
+
+}  // namespace FOEDAG
+
+#endif  // __TAP_H__
```
Comment (chernyee-rs) [inline src/Configuration/HardwareManager/Tap.h:36]: add empty line

File: src/Configuration/Programmer/CfgStatus.h [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+#define __CFGSTATUS_H__
+
+namespace FOEDAG {
+struct CfgStatus {
+  bool cfgDone;
+  bool cfgError;
+};
+
+}  // namespace FOEDAG
+#endif  // __CFGSTATUS_H__
```
Comment (chernyee-rs) [inline src/Configuration/Programmer/CfgStatus.h:32]: add empty line

File: src/Configuration/Programmer/Programmer.h [hunk 15]
```diff
@@ -283,4 +223,4 @@ int ProgramFlash(const Cable& cable, const Device& device,
 
 }  // namespace FOEDAG
 
-#endif
+#endif
```
Comment (chernyee-rs) [inline src/Configuration/Programmer/Programmer.h:226]: add empty line

File: src/Configuration/Programmer/ProgrammerTool.cpp [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+
+int ProgrammerTool::query_fpga_status(const Device& device,
+                                      CfgStatus& cfgStatus,
+                                      std::string& outputMessage) {
+  int statusCode = ProgrammerErrorCode::NoError;
+  statusCode = m_adapter->query_fpga_status(device, cfgStatus, outputMessage);
+  return statusCode;
+}
+
+}  // namespace FOEDAG
```
Comment (chernyee-rs) [inline src/Configuration/Programmer/ProgrammerTool.cpp:72]: add empty line

File: src/Configuration/Programmer/ProgrammerTool.h [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+  int query_fpga_status(const Device& device, CfgStatus& cfgStatus,
+                        std::string& outputMessage);
+
+ private:
+  ProgrammingAdapter* m_adapter;
+};
+
+}  // namespace FOEDAG
+
+#endif  // PROGRAMMERTOOL_H
```
Comment (chernyee-rs) [inline src/Configuration/Programmer/ProgrammerTool.h:40]: add empty line

File: src/Configuration/Programmer/Programmer_cmd.cpp [hunk 18]
```diff
@@ -19,140 +19,8 @@ You should have received a copy of the GNU General Public License
...
-  if (ret != NoError) {
-    CFG_POST_MSG("Error code = %d. %s", ret, GetErrorMessage(ret).c_str());
-    return ret;
-  }
-  CFG_POST_MSG("### GetFpgaStatus API testing after programming ###");
-  CFG_POST_MSG("Device cfgDone: %d", status.cfgDone);
-  CFG_POST_MSG("Device cfgError: %d", status.cfgError);
-
-  return 0;
 }
```
Comment (chernyee-rs) [inline src/Configuration/Programmer/Programmer_cmd.cpp:26]: add empty line

File: src/Configuration/Programmer/Programmer_error_code.h [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+  CableNotFound = -7,
+  CableNotSupported = -8,
+  DeviceNotFound = -9,
+  BitfileNotFound = -10,
+  OpenOCDExecutableNotFound = -11,
+  InvalidFlashSize = -12,
+  ParseFpgaStatusError = -13
+};
+
+}
```
Comment (chernyee-rs) [inline src/Configuration/Programmer/Programmer_error_code.h:21]: add empty line

File: src/Configuration/Programmer/Programmer_helper.h [hunk 20]
```diff
@@ -97,31 +65,10 @@ std::string buildCableDevicesAliasNameWithSpaceSeparatedString(
...
-                          std::string& outputMsg);
-
-std::vector<TapInfo> extractTapInfoList(const std::string& tapInfoString);
-int extractDeviceList(const std::string& deviceListString,
-                      std::vector<Device>& devices);
 CfgStatus extractStatus(const std::string& statusString, bool& statusFound);
-std::string transportTypeToString(TransportType transport);
 
-}  // namespace FOEDAG
+}  // namespace FOEDAG
```
Comment (chernyee-rs) [inline src/Configuration/Programmer/Programmer_helper.h:74]: add empty line

File: src/ProgrammerGui/ProgrammerSettingsWidget.cpp [hunk 21]
```diff
@@ -77,4 +83,4 @@ void ProgrammerSettingsWidget::apply() {
                       ui->spinBoxFreq->text());
 }
 
-}  // namespace FOEDAG
+}  // namespace FOEDAG
```
Comment (chernyee-rs) [inline src/ProgrammerGui/ProgrammerSettingsWidget.cpp:86]: add empty line

File: src/Configuration/HardwareManager/CMakeLists.txt (hunk 2 above)
Comment (guanyung-rs) [reply src/Configuration/HardwareManager/CMakeLists.txt:89]: image
when running the formatting ./.github/bin/run-clang-format.sh, it does not allow empty line at end of file. It clean up.

```
guanyung@Lenovo-RS-GY:~/dev/FOEDAG$ ./.github/bin/run-clang-format.sh 
== There were changes running the formatter ==
diff --git a/src/Configuration/HardwareManager/CMakeLists.txt b/src/Configuration/HardwareManager/CMakeLists.txt
index eaa6f2ea..edcd9bde 100644
--- a/src/Configuration/HardwareManager/CMakeLists.txt
+++ b/src/Configuration/HardwareManager/CMakeLists.txt
@@ -86,4 +86,4 @@ foreach(file ${SRC_H_INSTALL_LIST})
     FILES ${PROJECT_SOURCE_DIR}/${file}
     DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/foedag/HardwareManager
   )
-endforeach()
\ No newline at end of file
+endforeach()
Submodule third_party/gtkwave contains modified content
To locally fix, run .github/bin/run-clang-format.sh then commit and push. 
```

File: src/Configuration/CFGCommon/CFGArg.json (hunk 1 above)
Comment (guanyung-rs) [reply src/Configuration/CFGCommon/CFGArg.json:273]: the formatting tool is cleaning up empty line. It doesn't allow that

File: src/Configuration/Programmer/Programmer.cpp [hunk 22]
```diff
@@ -21,43 +21,48 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
...
+#include "Configuration/HardwareManager/OpenocdAdapter.h"
 #include "ProgrammerGuiInterface.h"
+#include "ProgrammerTool.h"
+#include "Programmer_error_code.h"
 #include "Programmer_helper.h"
-#include "libusb.h"
 #include "tcl.h"
 namespace FOEDAG {
 
 // openOCDPath used by library
```
Comment (chernyee-rs) [inline src/Configuration/Programmer/Programmer.cpp:39]: Do we still need this? We can get the openocd path via cmdarg->toolPath.string().

File: src/Configuration/Programmer/Programmer.cpp [hunk 23]
```diff
@@ -200,57 +218,45 @@ void programmer_entry(CFGCommon_ARG* cmdarg) {
     int status = 0;
     InitLibrary(openOcdExecPath.string());
```
Comment (chernyee-rs) [inline src/Configuration/Programmer/Programmer.cpp:219]: Do we still need to init openocd library? This should be handled in openocd adapter.

File: src/Configuration/HardwareManager/CMakeLists.txt (hunk 2 above)
Comment (chernyee-rs) [reply src/Configuration/HardwareManager/CMakeLists.txt:89]: This is git diff warning telling us there is no newline at end of file. The clang formatter does allow this.

**Review by ravic-rs — CHANGES_REQUESTED** [review]
@guanyung-rs please address GUI testing for refactoring the programmer code.
It's not clear if this was testing with GUI as well.

File: src/Configuration/Programmer/Programmer.cpp (hunk 22 above)
Comment (guanyung-rs) [reply src/Configuration/Programmer/Programmer.cpp:39]: the `openOcdExecPath` is being used in the C style API `ListDevices`, `ProgramFpga`, `ProgramOtp`. We still need this.

File: src/Configuration/Programmer/Programmer.cpp (hunk 23 above)
Comment (guanyung-rs) [reply src/Configuration/Programmer/Programmer.cpp:219]: this API setup the openocdExecPath whenever user want to use the C style programmer API such as I `ListDevices`, `ProgramFpga`, `ProgramOtp`.

**Comment by guanyung-rs** [comment]
> @guanyung-rs please address GUI testing for refactoring the programmer code. It's not clear if this was testing with GUI as well.

@ravic-rs yeah, the testing of the GUI is done. All looks good and behave exactly like I expected.
Screenshot 2023-12-27 105750
Screenshot 2023-12-27 105924

**Review by esther-rs — APPROVED** [review]

**Review by chernyee-rs — APPROVED** [review]

**Review by ravic-rs — APPROVED** [review]

**Comment by ravic-rs** [comment]
> > @guanyung-rs please address GUI testing for refactoring the programmer code. It's not clear if this was testing with GUI as well.
> 
> 
> 
> @ravic-rs yeah, the testing of the GUI is done. All looks good and behave exactly like I expected.
> 
> Screenshot 2023-12-27 105750
> 
> Screenshot 2023-12-27 105924
> 
> 
> 
> 

Please push this all the way to raptor.

**Pull request merged**